### PR TITLE
Fix decoding of headers with static value

### PR DESCRIPTION
**Pull request by waahm7**
*Description of changes:*
This PR fixes the following issues with the decoding logic. 

- If we are decoding a header with static length value like int32 using `s_read_header_value` function, sometimes, it will set the `header_value.variable_len_val` as an optimization to save allocation. Afterwards, when we call `aws_event_stream_header_value_as_int32` on the returned value, it will always try to parse `header_value.static_val` as int32 which is undefined behavior and we get garbage value. For headers with static_val, we should always set `header_value.static_val`. The optimization should only happen for variable length headers as only those require allocation.
- In `s_read_header_type`, the check to validate the header type is valid, did not include `AWS_EVENT_STREAM_HEADER_BOOL_TRUE`. I am not sure why, as we already have a default case which throws invalid header error if type is not valid. I have removed the redundant check so that `AWS_EVENT_STREAM_HEADER_BOOL_TRUE` is parsed as a valid header.
- If we parse a boolean header, we need to update the state to `s_headers_state`, otherwise, an exception is thrown due to invalid state.
- Trivial: added const in `aws_event_stream_message_init`
- Fix `s_reset_state`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: source/event_stream.c [hunk 1]
```diff
@@ -1111,55 +1109,53 @@ static int s_read_header_type(
...
+        case AWS_EVENT_STREAM_HEADER_STRING:
+        case AWS_EVENT_STREAM_HEADER_BYTE_BUF:
+            decoder->state = s_read_header_value_len;
+            break;
+        case AWS_EVENT_STREAM_HEADER_BOOL_FALSE:
+            current_header->header_value_len = 0;
+            current_header->header_value.static_val[0] = 0;
+            decoder->on_header(decoder, &decoder->prelude, current_header, decoder->user_context);
+            s_reset_header_state(decoder, 1);
+            decoder->state = s_headers_state;
```
Comment (waahm7) [inline source/event_stream.c:1124]: Git is showing all the lines as a diff, I have only removed the if and added this line here and in `AWS_EVENT_STREAM_HEADER_BOOL_TRUE` case.

**Review by bretambrose — APPROVED** [review]
Is there any way to write a test that would fail in the old, buggy impl and not fail now?

File: source/event_stream.c [hunk 2]
```diff
@@ -1111,55 +1109,53 @@ static int s_read_header_type(
...
+            break;
+        case AWS_EVENT_STREAM_HEADER_BOOL_TRUE:
+            current_header->header_value_len = 0;
+            current_header->header_value.static_val[0] = 1;
+            decoder->on_header(decoder, &decoder->prelude, current_header, decoder->user_context);
+            s_reset_header_state(decoder, 1);
+            decoder->state = s_headers_state;
```
Comment (graebm) [inline source/event_stream.c:1131]: how did this ever work before ?!?!

it looks like the old code would remain in the `s_read_header_type()` state forever after reading a bool. How did this ever pass any test, or succeed in any implementation?

File: source/event_stream.c [hunk 3]
```diff
@@ -1013,7 +1013,8 @@ static int s_read_header_value(
     size_t length_read = current_pos - decoder->current_header_value_offset;
     struct aws_event_stream_header_value_pair *current_header = &decoder->current_header;
 
-    if (!length_read) {
+    if (!length_read && (current_header->header_value_type == AWS_EVENT_STREAM_HEADER_BYTE_BUF ||
+                         current_header->header_value_type == AWS_EVENT_STREAM_HEADER_STRING)) {
         /* save an allocation, this can only happen if the data we were handed is larger than the length of the header
```
Comment (graebm) [inline source/event_stream.c:1018]: how did this ever work before?

the old code would end up reading an arbitrary memory address as the value of an integer.... how did that ever pass any test?

File: source/event_stream.c (hunk 2 above)
Comment (waahm7) [reply source/event_stream.c:1131]: I think it never worked before, and we only had test for string header.

Comment (graebm) [reply source/event_stream.c:1131]: discussed offline. the reason we didn't catch these bugs before is that the streaming decoder is only currently used by aws-sdk-cpp, by 2 services, and neither service sends headers in its response messages

File: source/event_stream.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (waahm7) [reply source/event_stream.c:1153]: I don't think a function named `s_reset_header_state` should also update decoder.

File: tests/message_streaming_decoder_test.c [hunk 4]
```diff
@@ -273,6 +285,113 @@ AWS_TEST_CASE(
     test_streaming_decoder_incoming_application_one_compressed_header_pair_valid,
     s_test_streaming_decoder_incoming_application_one_compressed_header_pair_valid_fn)
 
+static int s_test_streaming_decoder_incoming_application_one_int_header_pair_valid_fn(
```
[the remark this answers is not in the document]
Comment (waahm7) [reply tests/message_streaming_decoder_test.c]: Thanks, fixed.

File: tests/message_streaming_decoder_test.c [hunk 5]
```diff
@@ -273,6 +285,113 @@ AWS_TEST_CASE(
...
+    struct aws_allocator *allocator,
+    void *ctx) {
+    (void)ctx;
+    uint8_t test_data[] = {0x00, 0x00, 0x00, 0x1e, 0x00, 0x00, 0x00, 0x0e, 0x5d, 0x4a, 0xdb, 0x8d, 0x08, 0x65, 0x76,
+                           0x65, 0x6e, 0x74, 0x2d, 0x69, 0x64, 0x04, 0x00, 0x00, 0x00, 0x20, 0x2a, 0x4f, 0xb0, 0xea};
```
[the remark this answers is not in the document]
Comment (waahm7) [reply tests/message_streaming_decoder_test.c]: Thanks, fixed.

File: tests/message_streaming_decoder_test.c [hunk 6]
```diff
@@ -273,6 +285,113 @@ AWS_TEST_CASE(
...
+
+    ASSERT_BIN_ARRAYS_EQUALS(
+        expected_header_name,
+        strlen(expected_header_name),
+        decoder_data.latest_header_name,
+        strlen(decoder_data.latest_header_name),
+        "header name should have been %s",
+        expected_header_name);
+
+    int32_t latest_header_value = (int32_t)aws_read_u32(decoder_data.latest_header_value);
```
[the remark this answers is not in the document]
Comment (waahm7) [reply tests/message_streaming_decoder_test.c]: Thanks, I have added `aws_event_stream_add_header`

**Review by graebm — APPROVED** [review]

**Pull request merged**